### PR TITLE
Add end to end test for MS1 & Creation of an ArrowBack composable

### PR DESCRIPTION
**Pull request by Shayan105**
# Added Feature:
- Add an end-to-end test that goes over several screens of the app and check for all their components. 
- Modularization of the backward arrow used to go back in the navigation graph as it is used a lot.

# Target of the end-to-end test:
The target of the test is to make sure all the components of the app are reactive and displayed as they are meant to be.
The test workflow is described in the picture bellow. Here is the protocol : 
- When arriving on a screen, check that all components are displayed.
- If the navigation can go deeper, the test goes to the next screen other wise it goes back until it reached the main screen.

# Test limitations:
Unfortunately I was not able to use any of the cloud services due to a hard time trying to mock it. That's why the saving preferences of the user and the sign-in part are not tested. However there are already a plenty of test for these cases in the project.

image

### CODE REVIEW

File: app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  @Test
+  fun testEndToEnd() {
+    // Test begins at the main screen
+    // check that everything is displayed
+    mainScreenIsWellDisplayed()
+    // Go to the achievement screen
+    composeTestRule.onNodeWithTag("Achievements").performClick()
+    achievementScreenIsWellDisplayed()
+    // go back to the main screen
+    composeTestRule.onNodeWithTag("Main").performClick()
```
Comment (AlecLeGrooker) [inline app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt:43]: Check the new implementation of the main screen as some the test tags might have changed

File: app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    composeTestRule.onNodeWithTag("mainScreen").assertIsDisplayed()
```
Comment (AlecLeGrooker) [inline app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt:44]: Same for this tag

File: app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    // go to the video screen
+    composeTestRule.onNodeWithTag("Video").performClick()
+    composeTestRule.onNodeWithTag("videoScreen").assertIsDisplayed()
+    // go back to the main screen
+    composeTestRule.onNodeWithTag("Main").performClick()
+    composeTestRule.onNodeWithTag("mainScreen").assertIsDisplayed()
+    // go to the settings screen
+    composeTestRule.onNodeWithTag("SettingsButton").performClick()
```
Comment (AlecLeGrooker) [inline app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt:52]: Normally this tag didn't changed

File: app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    composeTestRule.onNodeWithTag("preferencesButton").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("deleteAccountButton").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("logoutButton").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("userDataButton").assertIsDisplayed()
+    // navigation bar should not be displayed
+    composeTestRule.onNodeWithTag("bottomNavigationMenu").assertDoesNotExist()
+  }
+
+  private fun mainScreenIsWellDisplayed() {
+    composeTestRule.onNodeWithTag("mainScreen").assertIsDisplayed()
```
Comment (AlecLeGrooker) [inline app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt:84]: Same for tag

File: app/src/main/java/com/android/sample/ui/mainscreen/MainScreen.kt [hunk 5]
```diff
@@ -25,12 +25,14 @@ fun MainScreen(navigationActions: NavigationActions) {
         TopAppBar(
             title = { Text("Main Screen") },
             actions = {
-              IconButton(onClick = { navigationActions.navigateTo("Settings Screen") }) {
-                Icon(
-                    imageVector = Icons.Outlined.Settings,
-                    contentDescription = "Settings",
-                    tint = Color.Black)
-              }
+              IconButton(
```
Comment (AlecLeGrooker) [inline app/src/main/java/com/android/sample/ui/mainscreen/MainScreen.kt:28]: Again the code I added on MainScreen is similar but you might want to take a look at it.

**Review by AlecLeGrooker — COMMENTED** [review]
Good work overall ! I didn't read everthing in detail, I let Nada do it but I wanted to indicate to you where some changes might be necessary because of a merge that modify the MainScreen.kt.

File: app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      throw Exception("logoutButton not displayed in settingScreenIsWellDisplayed")
+    }
+    if (composeTestRule.onNodeWithTag("userDataButton").isNotDisplayed()) {
+      throw Exception("userDataButton not displayed in settingScreenIsWellDisplayed")
+    }
+    // navigation bar should not be displayed
+    if (composeTestRule.onNodeWithTag("bottomNavigationMenu").isDisplayed()) {
+      throw Exception(
+          "bottomNavigationMenu should not be displayed in settingScreenIsWellDisplayed")
+    }
```
Comment (Siupahc) [inline app/src/androidTest/java/com/android/sample/endToend/EndToEndTest1.kt:86]: Interesting way to show where exactly the test fails on the CI

**Review by Siupahc — APPROVED** [review]
Overall lgtm.

Minor detail, but on the graph it's specified that the endToEnd test goes to WorkoutSelectionScreen, but it currently doesn't. Still, the test is good

**Pull request merged**